ql/src/java/org/apache/hadoop/hive/ql/ddl/table/storage/compact/AlterTableCompactOperation.java

```diff
@@ -108,7 +119,9 @@ public AlterTableCompactOperation(DDLOperationContext context, AlterTableCompact
       }
       // If Iceberg table had partition evolution, it will create compaction request without partition specification,
       // and it will compact all files from old partition specs, besides compacting partitions of current spec in parallel.
-      if (DDLUtils.isIcebergTable(table) && table.getStorageHandler().hasUndergonePartitionEvolution(table)) {
+      if (DDLUtils.isIcebergTable(table) && table.getStorageHandler().hasUndergonePartitionEvolution(table) && 
+          (desc.getFilterExpr() == null || !table.getStorageHandler()
+              .getPartitionsWithFilter(table, desc.getFilterExpr(), false).isEmpty())) {
```

Comment:
how is that different from `isFilterMatching`?

Reply:
I am using `isFilterMatching` (renamed to `hasDataMatchingFilterExpr` for better readability) for checking if filter matches any records on an unpartitioned table. 

The above piece of code checks if the Iceberg table had partition evolution and filter expression matches any records from old partition specs.